### PR TITLE
design doc for adding show train stmt to SQLFlow

### LINKED ISSUE #2049: Allow users to query model meta information
**Is your feature request related to a problem? Please describe.**
There's no good method for a user to know how a model is trained. If a user is tuning an application and train several models, it's hard for her to remember those hyperparameters.
**Describe the solution you'd like**
Support something like `DESCRIBE table1` or `SHOW CREATE TABLE table1`. For example:
```sql
SHOW TRAIN iris.dnn;
```
SQLFlow can output the original training statement as 
```
+--------+-------------------------------------+-------------------------------------+
| Model  | TO TRAIN                            | stdout                              |
+--------+-------------------------------------+-------------------------------------+
|iris.dnn| SELECT * FROM iris.train            |=== Result for training set: ======  |
|        | TO TRAIN DNNCLASSIFIER WITH         | loss: 0.850680483341217             |
|        |     ......                          | accuracy: 0.73256                   |
|        | INTO iris.dnn;                      |......                               |
+--------+-------------------------------------+-------------------------------------+
```
**Describe alternatives (optional)**
Require the users to hand-coding the hyperparameters in the model name.
**Additional Notes**
This was one of the original design details of the Model Zoo.

### PR DESCRIPTION
**Pull request by lhw362950217**
Fix #2049

### CODE REVIEW

File: doc/design/show_train_stmt.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  `sepal_length` float DEFAULT NULL,
+  `sepal_width` float DEFAULT NULL,
+  `petal_length` float DEFAULT NULL,
+  `petal_width` float DEFAULT NULL,
+  `class` int DEFAULT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
++-------+-----------------------------------------------------------------+
+```
+
+It worth for SQLFlow to do the same thing. On SQLFlow platform, user may create a lot of models with a minor set of parameters different with each other. After a while, the  user may hardly tell what the differences are if he hadn't record the information else where (like in model name).  We would like to provide a formal way to solve this problem. Here we introduce the SHOW TRAIN feature.
```
Comment (shendiaomo) [inline doc/design/show_train_stmt.md]: > After a while, the  user may hardly tell

                  ^There're two whitespaces here.

File: doc/design/show_train_stmt.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Grammar
+```
+SHOW TRAIN table_where_model_stored;
+```
+where
+- `table_where_model_stored` must be a table which stores the trained model in MySQL or Hive
```
Comment (shendiaomo) [inline doc/design/show_train_stmt.md]: The model can also be in a filesystem such as `OSS`

File: doc/design/show_train_stmt.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  `class` int DEFAULT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
++-------+-----------------------------------------------------------------+
+```
+
+It worth for SQLFlow to do the same thing. On SQLFlow platform, user may create a lot of models with a minor set of parameters different with each other. After a while, the  user may hardly tell what the differences are if he hadn't record the information else where (like in model name).  We would like to provide a formal way to solve this problem. Here we introduce the SHOW TRAIN feature.
+
+## Grammar
+```
+SHOW TRAIN table_where_model_stored;
```
Comment (typhoonzero) [inline doc/design/show_train_stmt.md]: The train clause use `TO TRAIN` so maybe we should use `SHOW TO TRAIN` here to be consistent.

File: doc/design/show_train_stmt.md (hunk 2 above)
Comment (typhoonzero) [reply doc/design/show_train_stmt.md]: When training on PAI and save the model on OSS, the original train SQL statement is not saved yet. Implementing this case, we need to update the current training code.

File: doc/design/show_train_stmt.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
++-------+------------------------------------------------------------------+
+| test  | SELECT * FROM train to TRAIN xgboost.gbtree with learning_rate=0.4,
+ objective=multi:softmax, num_class=3 LABEL class INTO test;
++-------+-----------------------------------------------------------------+
+```
+In the future we may support to show more metadata other than the original sql statement.
+
+## Implementation
+
+- Extend the SQLFlow parser with our `SHOW TRAIN` statement
```
Comment (typhoonzero) [inline doc/design/show_train_stmt.md]: How? The current parser implementation will call the dialect parser and getting an error position and parse the extended train statements. To parse `SHOW TRAIN ...` we may need to parse if the statement is `SHOW TRAIN ...` before calling the dialect parser.

File: doc/design/show_train_stmt.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+- Branch to execute a `ShowTrain` cmd in executor, this cmd will first read the model on a `sqlfs` and  then extract the original train statement
+- Format and route the cmd's output to the given output buffer
+
+## Discuss
+I noticed model zoo has created a table to store model meta in MySQL, shall we just use this mechanism to get the meta or extract from the saved model? How about the hive ones?
```
Comment (typhoonzero) [inline doc/design/show_train_stmt.md]: This table will be deleted. We should save model metadata together with the model, then when publishing the model to model zoo, the metadata will be published to the model zoo.

Related: https://github.com/sql-machine-learning/sqlflow/issues/1759

File: doc/design/show_train_stmt.md (hunk 3 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: As I see, SHOW TRAIN is more like a standard SQL stmt, which may lead to a more consistent user experience for analysts. So I still prefer not use a 'TO' here.

File: doc/design/show_train_stmt.md (hunk 4 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: Yes, as with this case, maybe we can do some hard-coded preprocessing to escape the dialect parser.
However, as we introduce more and more grammer features, we have to distinguish each kind of stmt sooner or later. Not all cases can fit the current `SELECT ... TO ...` pattern. We may need a more powerful customized parser to extract standard SQL and the extended SQL in a unified manner.

File: doc/design/show_train_stmt.md (hunk 5 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: roger that

File: doc/design/show_train_stmt.md (hunk 3 above)
Comment (typhoonzero) [reply doc/design/show_train_stmt.md]: Just for discussion, `SHOW CREATE TABLE [table_name]`  is showing statement `CREATE TABLE [table_name] ...`, yet `SHOW TRAIN [trained_model_name]` is showing statement `SELECT ... TO TRAIN ... INTO [trained_model_name]`. Maybe not straight forward, I'd prefer `DESCRIBE MODEL [trained_model_name]` to show all model metadatas like training statement as well as column information generated by feature derivation.

File: doc/design/show_train_stmt.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# Show Train Statement in SQLFlow
+
+## Background
+
+Like using other SQL engines, people may want to access the table metadata (say, the  table creating statement) using a SQL query. Usually they will type a statement like:
```
Comment (Yancey0623) [inline doc/design/show_train_stmt.md]: > the  table creating statement)
Redundant whitespace between `the` and `table`

Comment (Yancey0623) [inline doc/design/show_train_stmt.md]: > the  table creating statement

the `CREATE TABLE` statement

File: doc/design/show_train_stmt.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ objective=multi:softmax, num_class=3 LABEL class INTO test;
++-------+-----------------------------------------------------------------+
+```
+In the future we may support to show more metadata other than the original sql statement.
+
+## Implementation
+
+- Extend the SQLFlow parser with our `SHOW TRAIN` statement
+- Branch to execute a `ShowTrain` cmd in executor, this cmd will first read the model on a `sqlfs` and  then extract the original train statement
+- Format and route the cmd's output to the given output buffer
```
Comment (Yancey0623) [inline doc/design/show_train_stmt.md:86]: How about using pseudocode to describe the implementation?
I have the same question with @typhoonzero .

File: doc/design/show_train_stmt.md (hunk 3 above)
Comment (Yancey0623) [inline doc/design/show_train_stmt.md]: `SHOW` and `TRAIN` are all verb, it does not comply with grammar rules.
`SHOW MODEL` or `DESCRIBE MODEL` seems better?

File: doc/design/show_train_stmt.md (hunk 4 above)
Comment (tonyyang-svail) [reply doc/design/show_train_stmt.md]: Maybe mentioning some implementation details on "a more powerful customized parser".

File: doc/design/show_train_stmt.md (hunk 3 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: We may implement this feature later, as I see, current serialized model has no column info.

Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: We finally decided to use `SHOW TRAIN` for it more like SQL grammer`SHOW TRAIN TABLE` etc.

File: doc/design/show_train_stmt.md (hunk 7 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md:86]: Yes

File: doc/design/show_train_stmt.md [hunk 8]
```diff
@@ -22,31 +21,54 @@ mysql> SHOW CREATE TABLE `test`;
...
-- Branch to execute a `ShowTrain` cmd in executor, this cmd will first read the model on a `sqlfs` and  then extract the original train statement
+- Extend the SQLFlow parser with our `SHOW TRAIN` statement. First, we need to add a key word `SHOW` to our extended grammar. In addition, `SHOW TRAIN` is not like our train/predict/explain statements which all share a format `SELECT ... TO ...` in which there is a **standard** `SELECT` part at the front and an **extended** `TO ...` part at the end. With this definition, our extending statement has no standard part. So, we have to modify the parse process slightly. The pseudo code is like below:
+```go
+func Parse(program string) ([]*SQLFlowStmt, error) {
+  all := make([]*SQLFlowStmt)
+  for program != "" {
+    standard, err := tryExtractStandardPart(program)
+    if err == nil {
+      all = append(all, standard)
+    } else if extended, e := tryParseByExtendedGrammer(program); e == nil {
```
Comment (shendiaomo) [inline doc/design/show_train_stmt.md]: Grammer->Grammar

File: doc/design/show_train_stmt.md [hunk 9]
```diff
@@ -22,31 +21,54 @@ mysql> SHOW CREATE TABLE `test`;
...
+      }
+    } else {
+      return nil, err
+    }
+    program = eatConsumedPrefix(program)
+  }
+  return all, nil
+}
+```
+- Branch to execute a `ShowTrain` cmd in executor, this cmd will first read the model on a `sqlfs` and then extract the original train statement. We can not get the train statement from models saved on OSS because they have not been saved with the models yet. This can be implemented next time.
```
Comment (shendiaomo) [inline doc/design/show_train_stmt.md:85]: Please add an issue for putting the train statements into the OSS and update this document after fix the issue.

**Review by typhoonzero — DISMISSED** [review]
LGTM, please fix @shendiaomo 's comment before merge.

File: doc/design/show_train_stmt.md (hunk 9 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md:85]: Done, take a look at #2085

File: doc/design/show_train_stmt.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  `petal_length` float DEFAULT NULL,
+  `petal_width` float DEFAULT NULL,
+  `class` int DEFAULT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci |
++-------+-----------------------------------------------------------------+
+```
+
+It worth for SQLFlow to do the same thing. On SQLFlow platform, user may create a lot of models with a minor set of parameters different with each other. After a while, the user may hardly tell what the differences are if he hadn't record the information else where (like in model name). We would like to provide a formal way to solve this problem. Here we introduce the `SHOW TRAIN` feature.
+
+## Grammar
```
Comment (Yancey0623) [inline doc/design/show_train_stmt.md]: I prefer using `Syntax` instead of `Grammar`, because

> The syntax of a programming language is the set of all syntactically valid programs.
A grammar is one way to specify a formal language.

Grammar is context-free. ref: https://cs.stackexchange.com/questions/62765/relation-between-syntax-and-grammar-in-cs

Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: 👍

File: doc/design/show_train_stmt.md (hunk 6 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: 👍

File: doc/design/show_train_stmt.md (hunk 3 above)
Comment (lhw362950217) [reply doc/design/show_train_stmt.md]: There are two considerations. First, `SHOW CREATE` in SQL is to print the query that create this table, so, like that, we want `SHOW TRAIN` to print the model training query. Secondly, `DESCRIBE model` is a valid SQL query which print the colum meta. So if we use this form, it will be a prefix of our extended syntax, this may make our parsing process more complicated.

**Review by Yancey0623 — APPROVED** [review]
LGTM++, it's better to create an issue that contains the TODO tasks when we begin to develop this feature.

**Review by shendiaomo — APPROVED** [review]
LGTM++
